### PR TITLE
[WIP] add a preprocessor

### PR DESCRIPTION
**Pull request by jcmgray**
## Description

This adds a potential preprocessor that takes an equation and converts it into a minimal form with the following simplifications:

- All indices which only appear on a single input (and not the output) are summed over.
- All indices which appear multiple times on the same term are traced.
- All scalars are multiplied into the smallest other term
- All terms with the same indices are multiplied (hamadard product / elementwise) into a single term ('de-duplication').

along with a function, that transforms input arrays into the inputs for the new simplified `eq`. It is a function so that it could be part of a contract 'expression'.

This would address (#114, #99, #112, #167, #189, ...).

Example from the docstring:

```python
eq = ',ab,abee,,cd,cd,dd->ac'
arrays = helpers.build_views(eq)
new_eq, simplifier = make_simplifier(eq)
new_eq
# 'ab,cd,d->ac'
sarrays = simplifier(arrays)
oe.contract(new_eq, *sarrays)
# array([[0.65245661, 0.14684493, 0.42543411, 0.32350895],
#        [0.38357005, 0.08632807, 0.25010672, 0.19018636]])
```

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] when and how should this be called?
  - [ ] return a partial path rather than a 'simplifier'?
  - [ ] should shape information (`size_dict`) be used and propagated for use with a contraction expression (and so that the smallest tensor to multiply scalars into can be most accurately chosen)?
  - [ ] other simplifications are possible (e.g. hadamard deduplication *up to transpose*, performing *all* non-rank increasing contractions)...
  - [x] ~currently the parsing is a single sweep, should it iteratively sweep til no more changes? E.g. currently examples like `ab,ab->` transform to `ab->` but if processed again could transform to `->`~

## Status
- [ ] Ready to go

### CODE REVIEW

File: opt_einsum/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
+import operator
+import functools
+import collections
+
+from .backends import get_func
+from .parser import find_output_str
+
+
+def unique(it):
```
Comment (dgasmith) [inline opt_einsum/preprocessing.py:9]: Can we type these? I cannot quite tell what this is doing, somewhat looks like a set operation.

File: opt_einsum/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    def __repr__(self):
+        return (
+            f"SimplifyExpression('"
+            f"{','.join(lz.inds for lz in self._lazy_inputs)}->"
+            f"{','.join(lz.inds for lz in self._lazy_outputs)}')"
+        )
+
+
+def make_simplifier(*args, backend="numpy"):
```
Comment (dgasmith) [inline opt_einsum/preprocessing.py:68]:
```suggestion
def expression_simplifier(*args, backend="numpy"):
```

File: opt_einsum/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        >>> new_eq, simplifier = make_simplifier(eq)
+        >>> new_eq
+        'ab,cd,d->ac'
+
+        >>> sarrays = simplifier(arrays)
+        >>> oe.contract(new_eq, *sarrays)
+        array([[0.65245661, 0.14684493, 0.42543411, 0.32350895],
+               [0.38357005, 0.08632807, 0.25010672, 0.19018636]])
+    """
+    try:
```
Comment (dgasmith) [inline opt_einsum/preprocessing.py:112]: We will need quite a few tests around this tech, too bad we cannot reproduce the original parsing tech here.

**Review by dgasmith — COMMENTED** [review]
Makes sense, it would be good to start seeing some test here to find edge cases. A few notes:
- I would make a comment pass, see if we can highlight which operation we are reducing at each stage.
- The input/output pairing worries me a little bit. I wonder if we leave this as "internal" `oe.contract(..., preprocess=True)`
- It would be good to have a pathway for the `propocess=True` pathway.

Super cool addition!
